pandas/tests/frame/methods/test_replace.py

```diff
@@ -1469,20 +1468,23 @@ def test_regex_replace_scalar(
         tm.assert_frame_equal(result, expected)
 
     @pytest.mark.parametrize("regex", [False, True])
-    def test_replace_regex_dtype_frame(self, regex):
+    @pytest.mark.parametrize("value", [1, "1"])
+    def test_replace_regex_dtype_frame(self, regex, value):
         # GH-48644
         df1 = DataFrame({"A": ["0"], "B": ["0"]})
-        expected_df1 = DataFrame({"A": [1], "B": [1]}, dtype=object)
-        result_df1 = df1.replace(to_replace="0", value=1, regex=regex)
+        # When value is an integer, coerce result to object.
+        # When value is a string, infer the correct string dtype.
```

Comment:
Are we sure we want to coerce to string instead of raising? The object case makes sense I'm just not as sure onn the string side if we should be implicitly casting like that

Reply:
Not sure I understand. When `infer_string=True`, the input DataFrame is `str` dtype. Then when we go to replace `"0"` with `value="1"`, certainly we want the result to still be `str` dtype, no?

Reply:
Yea I think that makes sense, but I'm not as sure when the target value is a non-string, i.e. `replace(..., value=1)`

Reply:
When the target value is a non-string, we coerce to object dtype in order to hold both integers and strings. What are you not sure about?

Reply:
Sorry I was just misreading the comment - I think this is good